### PR TITLE
Unify HTTP token parsing code

### PR DESCRIPTION
**Pull request by adityamandaleeka**
While looking into some perf optimizations, I noticed that we had nearly identical code in two places, HttpRuleParser and HttpTokenParsingRule, used by different parts of the codebase.

It seems like HttpRuleParser has received some updates over time which weren't applied to HttpTokenParsingRule. The diffs looked innocuous to me, so this PR just moves the HttpRuleParser to shared source and includes it in the various projects that were using HttpTokenParsingRule.

The MVC core tests (which include tests that exercise the token parsing functionality) all passed for me. Let's see how CI goes 😄.

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Formatters/AcceptHeaderParser.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 using System.Diagnostics;
 using Microsoft.AspNetCore.Http.Headers;
```
Comment (Tratcher) [inline src/Mvc/Mvc.Core/src/Formatters/AcceptHeaderParser.cs:5]:
```suggestion
```

**Review by Tratcher — APPROVED** [review]

**Comment by javiercn** [comment]
@adityamandaleeka @Tratcher I don't fully remember the history, but at some point, we had our own parsing for the accept header because we were more lenient on accepting malformed accept headers. I believe Justin did some work to unify both of the parsers, but I don't remember the exact details. Just something to consider.

**Comment by amcasey** [comment]
> @adityamandaleeka @Tratcher I don't fully remember the history, but at some point, we had our own parsing for the accept header because we were more lenient on accepting malformed accept headers. I believe Justin did some work to unify both of the parsers, but I don't remember the exact details. Just something to consider.

I don't see any lenience differences - just extra functionality (and rephrased asserts).

However, I am curious how we ended up with code that was clearly copy-pasted.  Do we know the history of these files?  Were we just unable (or unaware that we could) use file linking?

**Comment by Tratcher** [comment]
MVC copied the code as part of the copying the MediaType class from MediaTypeHeaderValue because they wanted to tweak the functionality. They probably copied more than they needed.

**Comment by amcasey** [comment]
@Tratcher What did they tweak?

**Comment by adityamandaleeka** [comment]
> @adityamandaleeka @Tratcher I don't fully remember the history, but at some point, we had our own parsing for the accept header because we were more lenient on accepting malformed accept headers. I believe Justin did some work to unify both of the parsers, but I don't remember the exact details. Just something to consider.

Thanks for that context @javiercn. I didn't spot any such differences currently though (and I'm glad @amcasey checked it too).

**Pull request merged**